caom2-pkg-server/src/main/java/ca/nrc/cadc/caom2/pkg/Caom2PackageRunner.java

```diff
@@ -0,0 +1,272 @@
+import java.security.cert.CertificateException;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+import javax.security.auth.Subject;
+import org.apache.log4j.Logger;
+
+
+public class Caom2PackageRunner extends PackageRunner {
+    private static final Logger log = Logger.getLogger(Caom2PackageRunner.class);
+
+    private final URI tapID;
+    private List<String> idList;
+    private String packageName;
+
+    public Caom2PackageRunner()
+    { 
+        ServiceConfig sc = new ServiceConfig();
+        this.tapID = sc.getTapServiceID();
+    }
+
+    @Override
+    protected String getPackageName() {
+        return this.packageName;
+    }
+
+
+    @Override
+     public Iterator<PackageItem> getItems() throws IOException, PackageRunnerException {
+
+        // obtain credentials from CDP if the user is authorized
+        AccessControlContext accessControlContext = AccessController.getContext();
+        Subject subject = Subject.getSubject(accessControlContext);
+        AuthMethod authMethod = AuthenticationUtil.getAuthMethod(subject);
+
+        String runID = this.job.getID();
+        if (this.job.getRunID() != null) {
+            runID = this.job.getRunID();
+        }
+
+        List<String> idList = ParameterUtil.findParameterValues("ID", this.job.getParameterList());
+
+        CaomTapQuery query = new CaomTapQuery(tapID, runID);
+
+        CaomArtifactResolver artifactResolver = new CaomArtifactResolver();
+        artifactResolver.setAuthMethod(authMethod); // override auth method for proxied calls
+        artifactResolver.setRunID(runID);
+
+        List<PackageItem> packageItems = new ArrayList<PackageItem>();
+
+        for (String suri : idList) {
+
+            try {
+                // If this fails, the id will be skipped
+                URI uri = new URI(suri);
+
+                PlaneURI puri;
+                ArtifactQueryResult result;
+
+                // If either query fails, the id will be skipped
+                if (PublisherID.SCHEME.equals(uri.getScheme())) {
+                    PublisherID p = new PublisherID(uri);
+                    result = query.performQuery(p, true);
+                    puri = toPlaneURI(p);
+                } else {
+                    puri = new PlaneURI(uri);
+                    result = query.performQuery(puri, true);
+                }
+
+                List<Artifact> artifacts = result.getArtifacts();
+                stripPreviews(artifacts);
+
+                // set package name is done here because the plane URI is
+                // used to create it if there is only one
+                if (!StringUtil.hasText(packageName)) {
+                    if (idList.size() == 1) {
+                        // For a single id, package name is derived from
+                        // Publisher URI
+                        this.packageName = getFilenamefromURI(puri);
```

Comment:
hmmm. Unless the base PackageRunner guarentees that `getItems` is called before `getPackageName()` then this is too late and `getPackageName()` returned a null. There is no reason for the base class to call them in a specific order... we've had this kind of thing in other places (cadc-rest) and it can be a very subtle bug. 

The simplest thing is to have the `getPackageName()` method get the ID values from the job and figure it out rigyht there. So lines~177-188 should move into that method.

Alternatively, and this would effect the extension API, we could think about adding a method that was guarenteed to be called before the `get` methods (something like `protected void initPackage()`). In there, one could get the ID values, validate them all and populate a List<PlaneURI>, and figure out the package name. That method could throw IllegalArgumentException for invalid input. It might be overkill, but currently it's feasible to have an invalid ID that causes an exceptiolnl to be thrown during iteration (after output started), eg line 156, or 163, or 167... and that's not feasible to handled very well.

Reply:
Because the name depended on the number of IDs passed in which isn't discovered until getItems is called, and was derived from the publisher id /plane uri currently being processed, unless the ID list is processed outside getItems, the information just wasn't available. 

It could all be moved in (including the code to pull the uri from the idList and parse it into a single name) - 

The base PackageRunner has this code:
```

            // Build an iterator of PackageItem from the files named
            // in the local Job instance
            Iterator<PackageItem> packageItems = getItems();

            // TarWriter needs an output stream instantiated.
            initOutputStream(getPackageName());
            writer = new TarWriter(this.outputStreamCounter);

            while (packageItems.hasNext()) {
                PackageItem pi = packageItems.next();
                writer.write(pi);
            }
```
so yes, it does call getItems before getPackageName().  
     
But is the decision to have the initPackage() function?

Reply:
I see the comment now - it's above any of the 'green' (meaning a section of code to look at,) so I skipped over it.

Reply:
made initPackage() in PackageRunner, made a list of PublisherID as this was the simplest, and put all validation into initPackage().